drivers/sdhc/emmc_host.c

```diff
@@ -0,0 +1,1337 @@
+		host_io->bus_width = ios->bus_width;
+	}
+
+	/* Set HC signal voltage */
+	if (ios->signal_voltage != host_io->signal_voltage) {
+		LOG_DBG("signal_voltage: %d", ios->signal_voltage);
+		ret = emmc_set_voltage(emmc, ios->signal_voltage);
+		if (ret) {
+			LOG_ERR("Set signal volatge failed failed:%d", ret);
+			return ret;
+		}
+		host_io->signal_voltage = ios->signal_voltage;
+		k_msleep(10u);
+	}
+
+	/* Set card power */
+	if (host_io->power_mode != ios->power_mode) {
+		LOG_DBG("power_mode: %d", ios->power_mode);
+
+		ret = emmc_set_power(emmc, ios->power_mode);
+		if (ret) {
+			LOG_ERR("Set Bus Width failed:%d", ret);
+			return ret;
+		}
+		host_io->power_mode = ios->power_mode;
+	}
+
+	/* Set I/O timing */
+	if (host_io->timing != ios->timing) {
+		LOG_DBG("timing: %d", ios->timing);
+
+		ret = set_timing(emmc, ios->timing);
+		if (ret) {
+			LOG_ERR("Set timing failed:%d", ret);
+			return ret;
+		}
+		host_io->timing = ios->timing;
+	}
+
+	return ret;
+}
+
+static int emmc_get_card_present(const struct device *dev)
+{
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+
+	LOG_DBG("");
+
+	emmc->card_present = (bool)((regs->present_state >> 16u) & 1u);
+
+	if (!emmc->card_present) {
+		LOG_ERR("No MMC device detected");
+	}
+
+	return ((int)emmc->card_present);
+}
+
+static int emmc_execute_tuning(const struct device *dev)
+{
+#if defined(CONFIG_EMMC_HOST_TUNING)
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+
+	LOG_ERR("Tuning starting...");
+
+	regs->host_ctrl2 |= EMMC_HOST_START_TUNING;
+	while (!(regs->host_ctrl2 & EMMC_HOST_START_TUNING)) {
+		;
+	}
+
+	if (regs->host_ctrl2 & EMMC_HOST_TUNING_SUCCESS) {
+		LOG_ERR("Tuning Completed success");
+	} else {
+		LOG_ERR("Tuning failed");
+		return -EIO;
+	}
+#endif
+	return 0;
+}
+
+static int emmc_card_busy(const struct device *dev)
+{
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = emmc->regs;
+
+	LOG_DBG("");
+
+	if (regs->present_state & 7u) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int emmc_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	struct emmc_data *emmc = dev->data;
+	const struct emmc_config *config = dev->config;
+	volatile struct emmc_reg *regs = emmc->regs;
+	uint64_t cap = regs->capabilities;
+
+	LOG_DBG("");
+
+	memset(props, 0, sizeof(struct sdhc_host_props));
+	props->f_max = config->max_bus_freq;
+	props->f_min = config->min_bus_freq;
+	props->power_delay = config->power_delay_ms;
+
+	props->host_caps.vol_180_support = (bool)(cap & BIT(26u));
+	props->host_caps.vol_300_support = (bool)(cap & BIT(25u));
+	props->host_caps.vol_330_support = (bool)(bool)(cap & BIT(24u));
+	props->host_caps.suspend_res_support = false;
+	props->host_caps.sdma_support = (bool)(cap & BIT(22u));
+	props->host_caps.high_spd_support = (bool)(cap & BIT(21u));
+#if defined(CONFIG_EMMC_HOST_ADMA)
```

Comment:
Could this be a DT property of the hardware? It seems like this would be a property of the IP itself, not a software configuration. Alternatively, could this be probed at runtime since the driver uses PCIE?

Reply:
Make it DTS config is a good suggestion, but too many parameter via DTS was my concern.

Reply:
I'd prefer this to be a DTS setting over Kconfig

Reply:
updated to read from capabilities register.